File: app/src/main/java/com/example/dreamloaf/auth/RegisterFragment.java
package com.example.dreamloaf.auth;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;
import com.example.dreamloaf.R;
import com.example.dreamloaf.data.User;
import com.example.dreamloaf.databinding.FragmentRegisterBinding;
import com.example.dreamloaf.viewmodels.AuthViewModel;

public class RegisterFragment extends Fragment {
    private FragmentRegisterBinding binding;
    private AuthViewModel authViewModel;

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        binding = FragmentRegisterBinding.inflate(inflater, container, false);
        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        authViewModel = new ViewModelProvider(requireActivity()).get(AuthViewModel.class);

        setupLoginCheck();
        setupRegisterButton();
    }

    private void setupLoginCheck() {
        binding.etLogin.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                String login = s.toString().trim();
                if (!login.isEmpty()) {
                    checkLoginAvailability(login);
                }
            }
        });
    }

    private void checkLoginAvailability(String login) {
        authViewModel.checkLoginAvailable(login).observe(getViewLifecycleOwner(), isAvailable -> {
            if (!isAvailable) {
                binding.etLogin.setError("Это имя уже занято");
            } else {
                binding.etLogin.setError(null);
            }
        });
    }

    private void setupRegisterButton() {
        binding.btnRegister.setOnClickListener(v -> {
            String login = binding.etLogin.getText().toString().trim();
            String password = binding.etPassword.getText().toString().trim();

            if (binding.spRole.getSelectedItem() == null) {
                Toast.makeText(requireContext(), "Выберите роль", Toast.LENGTH_SHORT).show();
                return;
            }

            String role = binding.spRole.getSelectedItem().toString();

            if (login.isEmpty() || password.isEmpty()) {
                Toast.makeText(requireContext(), "Заполните все поля", Toast.LENGTH_SHORT).show();
                return;
            }

            User user = new User();
            user.setLogin(login);
            user.setPassword(password);
            user.setRole(role);

            authViewModel.registerUser(user).observe(getViewLifecycleOwner(), success -> {
                if (success) {
                    Toast.makeText(requireContext(),
                            "Регистрация успешна! Теперь войдите в систему",
                            Toast.LENGTH_SHORT).show();
                    Navigation.findNavController(v).navigate(
                            R.id.action_registerFragment_to_loginFragment);
                } else {
                    Toast.makeText(requireContext(),
                            "Ошибка регистрации. Имя пользователя занято",
                            Toast.LENGTH_SHORT).show();
                }
            });
        });
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        binding = null;
    }
}

File: app/src/main/java/com/example/dreamloaf/auth/RegisterFragment.kt
package com.example.dreamloaf.auth

import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.navigation.Navigation.findNavController
import com.example.dreamloaf.R
import com.example.dreamloaf.data.User
import com.example.dreamloaf.databinding.FragmentRegisterBinding
import com.example.dreamloaf.viewmodels.AuthViewModel

class RegisterFragment : androidx.fragment.app.Fragment() {
    private var binding: FragmentRegisterBinding? = null
    private var authViewModel: AuthViewModel? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentRegisterBinding.inflate(inflater, container, false)
        return binding!!.getRoot()
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        authViewModel =
            ViewModelProvider(requireActivity()).get<AuthViewModel>(AuthViewModel::class.java)

        setupLoginCheck()
        setupRegisterButton()
    }

    private fun setupLoginCheck() {
        binding!!.etLogin.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}

            override fun afterTextChanged(s: Editable) {
                val login = s.toString().trim { it <= ' ' }
                if (!login.isEmpty()) {
                    checkLoginAvailability(login)
                }
            }
        })
    }

    private fun checkLoginAvailability(login: String?) {
        authViewModel!!.checkLoginAvailable(login)
            .observe(getViewLifecycleOwner(), Observer { isAvailable: Boolean? ->
                if (isAvailable == false) {
                    binding!!.etLogin.setError("Это имя уже занято")
                } else {
                    binding!!.etLogin.setError(null)
                }
            })
    }

    private fun setupRegisterButton() {
        binding!!.btnRegister.setOnClickListener { v: View? ->
            val login = binding!!.etLogin.getText().toString().trim { it <= ' ' }
            val password = binding!!.etPassword.getText().toString().trim { it <= ' ' }

            if (binding!!.spRole.getSelectedItem() == null) {
                Toast.makeText(requireContext(), "Выберите роль", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            val role = binding!!.spRole.getSelectedItem().toString()

            if (login.isEmpty() || password.isEmpty()) {
                Toast.makeText(requireContext(), "Заполните все поля", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            val user = User()
            user.login = login
            user.password = password
            user.role = role
            authViewModel!!.registerUser(user)
                .observe(getViewLifecycleOwner(), Observer { success: Boolean? ->
                    if (success == true) {
                        Toast.makeText(
                            requireContext(),
                            "Регистрация успешна! Теперь войдите в систему",
                            Toast.LENGTH_SHORT
                        ).show()
                        findNavController(v!!).navigate(
                            R.id.action_registerFragment_to_loginFragment
                        )
                    } else {
                        Toast.makeText(
                            requireContext(),
                            "Ошибка регистрации. Имя пользователя занято",
                            Toast.LENGTH_SHORT
                        ).show()
                    }
                })
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }
}